.github/workflows/pr-gate.yml

```diff
@@ -0,0 +1,208 @@
+name: pr-gate
+
+on:
+  push:
+    branches: [ laserprec/ghaction-ci ]
+
+# This file defines following CI workflow:
+#
+#  ┌──────────┐   ┌─────────┐   ┌────────┐
+#  │  static  ├─┬─►  build* ├─┬─►  test  │
+#  │ analysis │ │ │  (cpu)  │ │ │ report │
+#  └──────────┘ │ └─────────┘ │ └────────┘         
+#               │ ┌─────────┐ │
+#               ├─►  build* ├─┤
+#               │ │ (spark) │ │
+#               │ └─────────┘ │
+#               │ ┌─────────┐ │
+#               └─►  build* ├─┘
+#                 │  (gpu)  │   <-- TODO: Coming Soon
+#                 └─────────┘
+#                    ....
+# *each runs in PARALLEL different combinations
+#  of python version, OS, test subsets, etc
+#
+#  ASCII chart created via https://asciiflow.com/
+
+jobs:
+###############################################
+############### STATIC-ANALYSIS ###############
+###############################################
+  static-analysis:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Python 3.6
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.6
+    
+    - name: Install dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+
+    - name: Run flake8
+      # TODO: re-enable this flake8 blocks (turned off to get a draft of the pipeline infrastracture)
+      continue-on-error: true
+      run: |
+       tox -e flake8
+
+###############################################
+################## CPU-BUILD ##################
+###############################################
+  build-cpu:
+    runs-on: ${{ matrix.os }}
+    needs: static-analysis
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python: [3.6]
+        # different kind of tests are located in tests/<unit|integration|smoke> folders
+        test-kind: ['unit'] 
+        # pytest markers configured in tox.ini. See https://docs.pytest.org/en/6.2.x/example/markers.html
+        test-marker: ['not gpu and not spark and not notebooks', 'not gpu and notebooks and not spark']
+
+    steps:
+    - uses: actions/checkout@v2
+    ################# Run Python tests #################
+    - name: Use Python ${{ matrix.python }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python }}
+    
+    - name: Install build dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+    
+    - name: Run ${{ matrix.test-kind }} tests ('${{ matrix.test-marker }}')
+      # '-e py' will use the default 'python' executable found in system path
+      # for why using tox, see: https://tox.readthedocs.io/en/latest/index.html
+      # tox will do:
+      #   1. build and install source distribution (sdist)
+      #   2. run static analysis on the code (not implemented yet)
+      #   3. run all of the specified test environment (i.e. run tests in different pyversions, etc)
```

Comment:
is running tests under different python versions managed by the github action matrix or tox configuration?

Reply:
Right now the versions are managed by github action matrix. Tox can certainty run multiple build environments in parallel in a single machine, but given that the CI agent has only 2 CPUs, it's better to distribute the tasks over multiple machines.